appserver/deployment/dol/src/main/java/com/sun/enterprise/deployment/util/ResourceValidator.java

```diff
@@ -0,0 +1,696 @@
+            action = "Create the necessary object before deploying the application.",
+            comment = "For the method validateJNDIRefs of com.sun.enterprise.deployment.util.ResourceValidator."
+    )
+    private static final String RESOURCE_REF_JNDI_LOOKUP_FAILED = "AS-DEPLOYMENT-00026";
+
+    String target;
+
+    DeploymentContext dc;
+
+    Application application;
+
+    @Inject
+    private Events events;
+
+    @Inject
+    Domain domain;
+
+    JNDINamespace myNamespace;
+    
+    public void postConstruct() {
+        events.register(this);
+    }
+
+    @Override
+    public void event(Event event) {
+        if (event.is(Deployment.AFTER_APPLICATION_CLASSLOADER_CREATION)) {
+            dc = (DeploymentContext) event.hook();
+            application = dc.getModuleMetaData(Application.class);
+            DeployCommandParameters commandParams = dc.getCommandParameters(DeployCommandParameters.class);
+            target = commandParams.target;
+            if (application == null)
+                return;
+            parseResources();
+            processResources();
+        }
+    }
+
+    /**
+     * Parse all the resources and store them in a namespace before starting the validation.
+     */
+    private void parseResources() {
+        myNamespace = new JNDINamespace();
+        parseResources(application);
+        for (BundleDescriptor bd : application.getBundleDescriptorsOfType(DOLUtils.warType())) {
+            parseResources((JndiNameEnvironment) bd);
+        }
+        for (BundleDescriptor bd : application.getBundleDescriptorsOfType(DOLUtils.carType())) {
+            parseResources((JndiNameEnvironment) bd);
+        }
+        for (BundleDescriptor bd : application.getBundleDescriptorsOfType(DOLUtils.ejbType())) {
+            parseResources((JndiNameEnvironment) bd);
+            EjbBundleDescriptor ebd = (EjbBundleDescriptor) bd;
+            for (EjbDescriptor ejb : ebd.getEjbs())
+                parseResources(ejb);
+        }
+
+        // Parse the Managed Beans
+        parseManagedBeans();
+
+        // Parse AppScoped resources
+        String appName = DOLUtils.getApplicationName(application);
+        Map<String, List<String>> resourcesList =
+                (Map<String, List<String>>) dc.getTransientAppMetadata().get(ResourceConstants.APP_SCOPED_RESOURCES_JNDI_NAMES);
+        myNamespace.storeAppScopedResources(resourcesList, appName);
+    }
+
+    private void parseManagedBeans() {
+        for (BundleDescriptor bd : application.getBundleDescriptors()) {
+            for (ManagedBeanDescriptor managedBean : bd.getManagedBeans()) {
+                myNamespace.store(managedBean.getGlobalJndiName(), (JndiNameEnvironment)bd);
+            }
+        }
+    }
+
+    private void parseResources(JndiNameEnvironment env) {
+        for (Object next : env.getResourceReferenceDescriptors()) {
+            parseResources((ResourceReferenceDescriptor) next, env);
```

Comment:
These are probably unnecessary since the three Bundle Descriptors we use(Web, EJB, App Client) generally return the correct class.